src/phongo_bson_encode.c

```diff
@@ -403,17 +421,15 @@ static void php_phongo_zval_to_bson_internal(zval* data, php_phongo_field_path*
 			if (instanceof_function(Z_OBJCE_P(data), php_phongo_document_ce)) {
 				php_phongo_document_t* intern = Z_DOCUMENT_OBJ_P(data);
 
-				bson_destroy(bson);
-				bson_copy_to(intern->bson, bson);
+				phongo_bson_copy_to_noinit(intern->bson, bson);
```

Comment:
Note that only `php_phongo_zval_to_bson_internal` was affected.

While `php_phongo_bson_append_object` also handles direct copying of Document and PackedArray data, it does so via `bson_append_document` and `bson_append_array()` functions, respectively. There was no problem there.

Reply:
Ah. I assumed that I could destroy the existing `bson_t` instance and have `bson_copy_to` handle the rest. If I understand this correctly, the issue is that `bson_destroy(bson)` frees the memory `bson` is pointing at. `bson_copy_to` doesn't catch this as it asserts on `dst`, which is still a valid `bson_t*` (i.e. not `NULL`) but pointing at memory no longer owned. Classic use of a footgun provided by C I have to admit, and a great catch on your part!

Reply:
That's part of the problem, but removing `bson_destroy()` also doesn't help us. The crux of the matter is that `bson_copy_to()` isn't designed to work on an allocated `bson_t` struct, which always exist in an initialized state.

I spoke to @kevinAlbs about this and we think it's technically possible to alter `bson_copy_to()` to support both, but it doesn't seem necessary at this time -- and a local copy method in PHPC works just fine for our needs.